src/js/assert.js

```diff
@@ -97,19 +96,76 @@ function strictEqual(actual, expected, message) {
   }
 }
 
+
+function strictDeepEqual(val1, val2) {
+  if (util.isNumber(val1)) {
+    if (util.isNumber(val2)) {
+      return (Number.isNaN(val1) && Number.isNaN(val2)) || (val1 === val2);
+    } else {
+      return false;
+    }
+  }
+
+  if (Array.isArray(val1)) {
+    if (Array.isArray(val2) && val1.length === val2.length) {
+      var i;
+      for (i = 0; i < val1.length; i++) {
+        if (!strictDeepEqual(val1[i], val2[i])) {
+          return false;
+        }
+      }
+
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  if (util.isDate(val1) && util.isDate(val2)) {
+    return val1.getTime() === val2.getTime();
+  }
+
+  if (util.isRegExp(val1) && util.isRegExp(val1)) {
+    return val1.source === val2.source && val1.flags === val2.flags;
+  }
+
+  if (util.isError(val1) && util.isError(val2)) {
+    return val1.message !== val2.message;
+  }
+
+  var aKeys = Object.keys(val1);
+
+  if (aKeys.length !== Object.keys(val2).length) {
+    return false;
+  }
+
+  for (i = 0; i < aKeys.length; i++) {
+    var key = aKeys[i];
+    if (!val2.hasOwnProperty(key) || val1[key] !== val2[key]) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+
+function isDeepStrictEqual(val1, val2) {
+  if (val1 && val1 === val2) {
+    return true;
+  }
+
+  return strictDeepEqual(val1, val2);
+}
+
+
 function deepStrictEqual(actual, expected, message) {
   if (typeof expected !== 'object') {
     return strictEqual(actual, expected, message);
   }
 
-  strictEqual(
-    Object.keys(actual).length,
-    Object.keys(expected).length,
-    message
-  );
-
-  for (var key in expected) {
-    deepStrictEqual(actual[key], expected[key], message);
+  if (!isDeepStrictEqual(actual, expected)) {
+    fail(actual, expected, message, '===');
   }
```

Comment:
@Txiaozhe Could you please run test once locally? You seems changed the behavior of this function.

Reply:
I run the `npm test`, get this error: 
```shell
IoT.js Build Succeeded!!

==> Build test dependencies

==> Build NAPI test module with /Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp

/Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp configure

/Users/jimtang/GitHub/ShadowNode/node_modules/.bin/node-gyp build

  CC(target) Release/obj.target/napi_construct/napi_construct.o
../napi_construct.c:35:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int]
NAPI_MODULE_INIT() {
^
../napi_construct.c:37:39: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
                                      ^
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
            ^
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
../napi_construct.c:37:13: error: use of undeclared identifier 'env'
../napi_construct.c:37:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_create_function(env, "Constructor", NAPI_AUTO_LENGTH,
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:39:40: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
                                       ^
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
            ^
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
../napi_construct.c:39:13: error: use of undeclared identifier 'env'
../napi_construct.c:39:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_create_reference(env, construct, 1, &ConstructRef));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:40:44: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
                                           ^
../napi_construct.c:40:58: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
                                                         ^
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
            ^
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
../napi_construct.c:40:13: error: use of undeclared identifier 'env'
../napi_construct.c:40:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  NAPI_CALL(env, napi_add_env_cleanup_hook(env, cleanup, env));
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:47:64: note: expanded from macro 'NAPI_CALL'
#define NAPI_CALL(env, the_call) NAPI_CALL_BASE(env, the_call, NULL)
                                                               ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../common.h:42:14: note: expanded from macro 'NAPI_CALL_BASE'
      return ret_val;                          \
             ^~~~~~~
../napi_construct.c:42:20: error: use of undeclared identifier 'env'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                   ^
../napi_construct.c:42:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:65:14: note: expanded from macro 'SET_NAMED_METHOD'
      return NULL;                                                   \
             ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../napi_construct.c:42:20: error: use of undeclared identifier 'env'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                   ^
../napi_construct.c:42:25: error: use of undeclared identifier 'exports'
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
                        ^
../napi_construct.c:42:3: warning: incompatible pointer to integer conversion returning 'void *' from a function with result type 'int' [-Wint-conversion]
  SET_NAMED_METHOD(env, exports, "Constructor", Constructor);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../common.h:69:14: note: expanded from macro 'SET_NAMED_METHOD'
      return NULL;                                                   \
             ^~~~
/Library/Developer/CommandLineTools/usr/lib/clang/9.0.0/include/stddef.h:105:16: note: expanded from macro 'NULL'
#  define NULL ((void*)0)
               ^~~~~~~~~~
../napi_construct.c:43:10: error: use of undeclared identifier 'exports'
  return exports;
         ^
6 warnings and 17 errors generated.
make: *** [Release/obj.target/napi_construct/napi_construct.o] Error 1
```
what happened about **napi**? Awkward so mush!

Reply:
These issues may caused by not compatible Node.js version. Try latest LTS version like v10.13 or newer and run tests again.

Reply:
Maybe I found the error. we have not `isNaN()` function in Number or util?
```js
'use strict';

var util = require('util');

console.log(util.isNaN(3));
console.log(Number.isNaN(3));
```
output:
```shell
TypeError: Expected a function.
    at undefined (/Users/jimtang/GitHub/ShadowNode/number.js:1:1)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:266:36)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:214:23)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/module.js.wrapped:301:26)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/iotjs.js.wrapped:18:2)
    at undefined (/Users/jimtang/GitHub/ShadowNode/src/js/iotjs.js.wrapped:1:1)
```